Support single-expression functions

sgade opened the pull request:
<!-- Thank you for your contribution to Gryphon! Remember it's OK to ask for help if you need it. -->

### What's in this pull request?

With this change, the following translation is done:

```swift
func example() -> String {
  "Hello, World"
}
```

```kotlin
fun example(): String = "Hello, World"
```

The translation is only done when the function has a return type that is
equal to the value of the expression. So there are the two cases:

```swift
func example2() -> String {
  example()
}

func example3() {
  example()
}
```

```kotlin
fun example2(): String = example()

fun example3() {
  example()
}
```

Due to changes in Swift 5.1, functions with a single expression and a return type
can omit the 'return' keyword. 

>  A closure that consists of only a single expression is understood to return the value of that expression.

Kotlin calls these "single-expression functions". See the documentation for more
info: https://kotlinlang.org/docs/reference/functions.html#single-expression-functions.

### Does this resolve an open issue?
Yeah, it resolves #72.
<!-- (Forgot your issue's number? look it up [here](https://github.com/vinivendra/Gryphon/issues)). -->

### Checklist for submitting a pull request:

- [x] Your code builds without any errors or warnings (warnings when running `prepareForBootstrapTests.sh` are OK).
- [x] You ran the unit tests and Bootstrapping tests for your platform.
  - [x] If you changed any `.kt` files in the `Test cases` folder, you also ran the Acceptance tests.
  <!-- - [ ] Optional: if you're on macOS, you also ran the tests on a Docker container (it's OK if you didn't). -->
- [x] You have added new tests that check your changes (if possible).
- [x] This pull request is targeting the `development` branch (if you're contributing code) or the `gh-pages` branch (if you're contributing to the website).

sgade commented:
I am unsure on how to add tests for this case.

Creating a single-expression function in the `Test cases/functions.swift` file did not do anything on `runTests.sh` calls. Appreciate any help!

vinivendra commented:
Thanks for the PR!

> Creating a single-expression function in the `Test cases/functions.swift` file did not do anything on `runTests.sh` calls.

Do you mean the `Test cases/functionCalls.swift` file? If you added a function there (but not in the `functionCalls.kt` file) there should be an error on the IntegrationTest. That's what happened for me when I cloned your changes.

If you created a new `Test cases/functions.swift` file, then you'd need to add it to the `testCases` list at the end of the `SharedTestUtilities.swift` file, so the tests know it exists. That said, I think adding this case to `functionCalls.swift` would be easier (it's already not just about "calls" anymore, I should really change that name 😅).

vinivendra commented on Sources/GryphonLib/KotlinTranslator.swift:
```diff
@@ -655,6 +655,8 @@ public class KotlinTranslator {
 		let parameterStrings = try functionDeclaration.parameters
 			.map { try translateFunctionDeclarationParameter($0, withIndentation: indentation) }
 
+			let singleExpressionStatement = self.isSingleExpressionFunction(functionDeclaration.statements, returnTypeString: returnTypeString)
```
Nitpick: is this indented one extra level?

sgade replied:
You're right. I was fixing my indentation (because I'm using spaces normally) and that must have slipped through.

vinivendra commented on Sources/GryphonLib/KotlinTranslator.swift:
```diff
@@ -697,7 +704,11 @@ public class KotlinTranslator {
+				result.append("\(indentation){\n")
+			} else {
+				result.append(" = ")
+			}
```
This might look a bit weird when translating a function that gets broken up into multiple lines. We should probably use something like `"\(indentation)= "` here.

It would also be good to add a test for that case.

sgade replied:
I actually changed it to `result.append("\( increaseIndentation(indentation) )= ")` as this aligns it more with the previous lines. Looking at your [example from below](https://github.com/vinivendra/Gryphon/pull/75#discussion_r451591801), this then results in:

```
            open fun aaaaaaaaaaaaa(
                bbbbbbbbbbbbbbbb: Int,
                ccccccccccccc: Int,
                ddddddddddddddddddddddddd: Int,
                eeeeeeeeeee: Int)
                : String
                = "abc"
```

I think this looks nice enough, given the input of the user.

vinivendra commented on Sources/GryphonLib/KotlinTranslator.swift:
```diff
@@ -730,16 +741,43 @@ public class KotlinTranslator {
 			result.append(try translateSubtrees(
 				statements,
-				withIndentation: indentation,
+				withIndentation: !singleExpressionStatement ? indentation : "",
```
This could cause problems if the single expression gets broken up into multiple lines (e.g. a closure or a long function call), since its indentation would have been reset. For instance, try translating this:

```` swift
class A {
	class B {
		class C {
			func aaaaaaaaaaaaa(bbbbbbbbbbbbbbbb: Int, ccccccccccccc: Int, ddddddddddddddddddddddddd: Int, eeeeeeeeeee: Int) -> String {
				"abc"
			}
			
			func d() -> String {
				aaaaaaaaaaaaa(bbbbbbbbbbbbbbbb: 0, ccccccccccccc: 0, ddddddddddddddddddddddddd: 0, eeeeeeeeeee: 0)
			}
		}
	}
}
````

It's a silly example, but it demonstrates the two problems I was talking about.

You could solve this by making `isSingleExpressionFunction` return the single expression itself (or nil). Then you could check if the expression is nil when adding the `{`s, and use `translateExpression` on it here with the correct indentation.

sgade replied:
You really know your stuff, thanks for that suggestion.
This looks way nicer now!

```kotlin
internal open class A {
    open class B {
        open class C {
            open fun aaaaaaaaaaaaa(
                bbbbbbbbbbbbbbbb: Int,
                ccccccccccccc: Int,
                ddddddddddddddddddddddddd: Int,
                eeeeeeeeeee: Int)
                : String
                = "abc"

            open fun d(): String = aaaaaaaaaaaaa(
                    bbbbbbbbbbbbbbbb = 0,
                    ccccccccccccc = 0,
                    ddddddddddddddddddddddddd = 0,
                    eeeeeeeeeee = 0)
        }
    }
}
```

vinivendra commented on Sources/GryphonLib/KotlinTranslator.swift:
```diff
@@ -730,16 +741,43 @@ public class KotlinTranslator {
+				return false
+		}
+
+		let subtrees = statement.printableSubtrees.compactMap { $0 }
```
I don't think you need `printableSubtrees` here, they're mainly for output. You could do something like `guard let expressionStatement = statement as? ExpressionStatement` above, and then just use `expressionStatement.expression` below.

sgade replied:
Yes, you are absolutely right. 👍

sgade commented:
Thanks for your review.
I forgot to mention, I rebased and force-pushed to have fewer WIP-commits in the tree after you merge this PR.

I also added one test in the way you described above.
I must have done something different the first time, but sure enough the tests failed after adding the `definition5`. So there you go.

vinivendra commented:
This is looking great, I'll take another look in a second. Could you also add tests for cases that break into multiple lines? You can copy my example or create your own, whatever you prefer.

vinivendra commented:
Yeah that's it, the rest looks good. If you don't have time to add the test today, I'll merge it like it is and we can add it later.

sgade commented:
I'm already on it.

sgade commented:
Done. Ran the full test to be sure. Integration tests pass.

I was having a lot of fun with a _really_ long and complicated function name, however, that turned out to be too unreadable even for testing purposes. So I stuck with your example.

vinivendra commented:
I actually laughed out loud reading the example 😆
Great work, thanks!
